File: templates/Enterprise-Template/src/typescript/enterprise-bot/src/dialogs/onboarding/onboardingDialog.ts
// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT License

import {
    StatePropertyAccessor,
    TurnContext } from 'botbuilder';
import {
    DialogTurnResult,
    TextPrompt,
    WaterfallDialog,
    WaterfallStepContext } from 'botbuilder-dialogs';
import { BotServices } from '../../botServices';
import { EnterpriseDialog } from '../shared/enterpriseDialog';
import { OnboardingResponses } from './onboardingResponses';
import { IOnboardingState } from './onboardingState';

export class OnboardingDialog extends EnterpriseDialog {

    // Fields
    private static readonly RESPONDER: OnboardingResponses = new OnboardingResponses();
    private readonly ACCESSOR: StatePropertyAccessor<IOnboardingState>;
    private STATE!: IOnboardingState;
    private DIALOG_IDS: DialogIds = new DialogIds();

    constructor(botServices: BotServices, accessor: StatePropertyAccessor<IOnboardingState>) {
        super(botServices, OnboardingDialog.name);

        this.ACCESSOR = accessor;
        this.initialDialogId = OnboardingDialog.name;

        const onboarding: ((sc: WaterfallStepContext<IOnboardingState>) => Promise<DialogTurnResult<any>>)[] = [
            this.askForName.bind(this),
            this.askForEmail.bind(this),
            this.askForLocation.bind(this),
            this.finishOnboardingDialog.bind(this)
        ];

        this.addDialog(new WaterfallDialog<IOnboardingState>(this.initialDialogId, onboarding));
        this.addDialog(new TextPrompt(this.DIALOG_IDS.NAME_PROMPT));
        this.addDialog(new TextPrompt(this.DIALOG_IDS.EMAIL_PROMPT));
        this.addDialog(new TextPrompt(this.DIALOG_IDS.LOCATION_PROMPT));
    }

    public async askForName(sc: WaterfallStepContext<IOnboardingState>): Promise<DialogTurnResult> {
        this.STATE = await this.getStateFromAccessor(sc.context);

        if (this.STATE.name) {
            return sc.next(this.STATE.name);
        } else {
            return sc.prompt(this.DIALOG_IDS.NAME_PROMPT, {
                prompt: await OnboardingDialog.RESPONDER.renderTemplate(
                    sc.context,
                    OnboardingResponses.RESPONSE_IDS.NamePrompt,
                    <string> sc.context.activity.locale)
            });
        }
    }

    public async askForEmail(sc: WaterfallStepContext<IOnboardingState>): Promise<DialogTurnResult> {
        this.STATE = await this.getStateFromAccessor(sc.context);
        this.STATE.name = sc.result;

        await OnboardingDialog.RESPONDER.replyWith(sc.context, OnboardingResponses.RESPONSE_IDS.HaveNameMessage, { name: this.STATE.name });

        return sc.prompt(this.DIALOG_IDS.EMAIL_PROMPT, {
            prompt: await OnboardingDialog.RESPONDER.renderTemplate(
                sc.context,
                OnboardingResponses.RESPONSE_IDS.EmailPrompt,
                <string> sc.context.activity.locale)
        });
    }

    public async askForLocation(sc: WaterfallStepContext<IOnboardingState>): Promise<DialogTurnResult> {
        this.STATE = await this.getStateFromAccessor(sc.context);
        this.STATE.email = sc.result;

        await OnboardingDialog.RESPONDER.replyWith(
            sc.context, OnboardingResponses.RESPONSE_IDS.HaveEmailMessage, { email: this.STATE.email });

        return sc.prompt(this.DIALOG_IDS.LOCATION_PROMPT, {
            prompt: await OnboardingDialog.RESPONDER.renderTemplate(
                sc.context,
                OnboardingResponses.RESPONSE_IDS.LocationPrompt,
                <string> sc.context.activity.locale)
        });
    }

    public async finishOnboardingDialog(sc: WaterfallStepContext<IOnboardingState>): Promise<DialogTurnResult> {
        this.STATE = await this.getStateFromAccessor(sc.context);
        this.STATE.location = <string> sc.result;

        await OnboardingDialog.RESPONDER.replyWith(
            sc.context,
            OnboardingResponses.RESPONSE_IDS.HaveLocationMessage,
            {
                name: this.STATE.name,
                location: this.STATE.location
            });

        return sc.endDialog();
    }

    private async getStateFromAccessor(context: TurnContext): Promise<IOnboardingState>  {
        const state: IOnboardingState | undefined = await this.ACCESSOR.get(context);
        if (!state) {
            const newState: IOnboardingState = {
                email: '',
                location: '',
                name: ''
            };
            await this.ACCESSOR.set(context, newState);

            return newState;
        }

        return state;
    }
}

class DialogIds {
    public NAME_PROMPT: string = 'namePrompt';
    public EMAIL_PROMPT: string = 'emailPrompt';
    public LOCATION_PROMPT: string =  'locationPrompt';
}

File: templates/Enterprise-Template/src/csharp/EnterpriseBotSample/EnterpriseBotSample/Dialogs/Onboarding/OnboardingDialog.cs
// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT License.

using System.Threading;
using System.Threading.Tasks;
using EnterpriseBotSample.Dialogs.Shared;
using Microsoft.ApplicationInsights;
using Microsoft.Bot.Builder;
using Microsoft.Bot.Builder.Dialogs;

namespace EnterpriseBotSample.Dialogs.Onboarding
{
    public class OnboardingDialog : EnterpriseDialog
    {
        private static OnboardingResponses _responder = new OnboardingResponses();
        private IStatePropertyAccessor<OnboardingState> _accessor;
        private OnboardingState _state;

        public OnboardingDialog(BotServices botServices, IStatePropertyAccessor<OnboardingState> accessor, IBotTelemetryClient telemetryClient)
            : base(botServices, nameof(OnboardingDialog))
        {
            _accessor = accessor;
            InitialDialogId = nameof(OnboardingDialog);

            var onboarding = new WaterfallStep[]
            {
                AskForName,
                AskForEmail,
                AskForLocation,
                FinishOnboardingDialog,
            };

            // To capture built-in waterfall dialog telemetry, set the telemetry client 
            // to the new waterfall dialog and add it to the component dialog
            TelemetryClient = telemetryClient;
            AddDialog(new WaterfallDialog(InitialDialogId, onboarding) { TelemetryClient = telemetryClient });
            AddDialog(new TextPrompt(DialogIds.NamePrompt));
            AddDialog(new TextPrompt(DialogIds.EmailPrompt));
            AddDialog(new TextPrompt(DialogIds.LocationPrompt));
        }

        public async Task<DialogTurnResult> AskForName(WaterfallStepContext sc, CancellationToken cancellationToken)
        {
            _state = await _accessor.GetAsync(sc.Context, () => new OnboardingState());

            if (!string.IsNullOrEmpty(_state.Name))
            {
                return await sc.NextAsync(_state.Name);
            }
            else
            {
                return await sc.PromptAsync(DialogIds.NamePrompt, new PromptOptions()
                {
                    Prompt = await _responder.RenderTemplate(sc.Context, sc.Context.Activity.Locale, OnboardingResponses.ResponseIds.NamePrompt),
                });
            }
        }

        public async Task<DialogTurnResult> AskForEmail(WaterfallStepContext sc, CancellationToken cancellationToken)
        {
            _state = await _accessor.GetAsync(sc.Context, () => new OnboardingState());
            var name = _state.Name = (string)sc.Result;

            await _responder.ReplyWith(sc.Context, OnboardingResponses.ResponseIds.HaveNameMessage, new { name });

            return await sc.PromptAsync(DialogIds.EmailPrompt, new PromptOptions()
            {
                Prompt = await _responder.RenderTemplate(sc.Context, sc.Context.Activity.Locale, OnboardingResponses.ResponseIds.EmailPrompt),
            });
        }

        public async Task<DialogTurnResult> AskForLocation(WaterfallStepContext sc, CancellationToken cancellationToken)
        {
            _state = await _accessor.GetAsync(sc.Context, () => new OnboardingState());
            var email = _state.Email = (string)sc.Result;

            await _responder.ReplyWith(sc.Context, OnboardingResponses.ResponseIds.HaveEmailMessage, new { email });

            return await sc.PromptAsync(DialogIds.LocationPrompt, new PromptOptions()
            {
                Prompt = await _responder.RenderTemplate(sc.Context, sc.Context.Activity.Locale, OnboardingResponses.ResponseIds.LocationPrompt),
            });
        }

        public async Task<DialogTurnResult> FinishOnboardingDialog(WaterfallStepContext sc, CancellationToken cancellationToken)
        {
            _state = await _accessor.GetAsync(sc.Context, () => new OnboardingState());
            _state.Location = (string)sc.Result;

            await _responder.ReplyWith(sc.Context, OnboardingResponses.ResponseIds.HaveLocationMessage, new { _state.Name, _state.Location });
            return await sc.EndDialogAsync();
        }

        private class DialogIds
        {
            public const string NamePrompt = "namePrompt";
            public const string EmailPrompt = "emailPrompt";
            public const string LocationPrompt = "locationPrompt";
        }
    }
}
